chain/indexer/manager.go

```diff
@@ -0,0 +1,204 @@
+	i.pool.Submit(func() {
+		atomic.AddInt64(&i.active, 1)
+		defer func() {
+			atomic.AddInt64(&i.active, -1)
+			span.End()
+		}()
+
+		ts := ts
+		success, err := i.TipSet(ctx, ts)
+		if err != nil {
+			log.Errorw("index manager suffered fatal error", "error", err, "height", ts.Height(), "tipset", ts.Key().String())
+			i.fatal = err
+			return
+		}
+		if !success {
+			log.Warnw("index manager failed to fully index tipset", "height", ts.Height(), "tipset", ts.Key().String())
+		}
+	})
+	return nil
+}
+
+// TipSet synchronously indexes and persists `ts`. TipSet returns an error if the Manager's Indexer encounters a
+// fatal error. TipSet returns false if one or more of the Indexer's tasks complete with a status `ERROR` or `SKIPPED`, else returns true.
+// Upon cancellation of `ctx` TipSet will persist all incomplete tasks with status `SKIPPED` before returning.
+func (i *Manager) TipSet(ctx context.Context, ts *types.TipSet) (bool, error) {
+	ctx, span := otel.Tracer("").Start(ctx, "Manager.TipSet")
+	defer span.End()
+	log.Infow("index tipset", "height", ts.Height())
+
+	var cancel func()
+	var procCtx context.Context // cancellable context for the task
+	if i.window > 0 {
+		// Do as much indexing as possible in the specified time window (usually one epoch when following head of chain)
+		// Anything not completed in that time will be marked as incomplete
+		procCtx, cancel = context.WithTimeout(ctx, i.window)
+	} else {
+		// Ensure all goroutines are stopped when we exit
+		procCtx, cancel = context.WithCancel(ctx)
+	}
+	defer cancel()
+
+	// asynchronously begin indexing tipset `ts`, returning results as they become avaiable.
+	taskResults, taskErrors, err := i.indexer.TipSet(procCtx, ts)
+	// indexer suffered fatal error, abort.
+	if err != nil {
+		return false, err
+	}
+
+	// there are no results, bail.
+	if taskResults == nil {
+		return true, nil
+	}
+
+	success := true
+	ctx, exportSpan := otel.Tracer("").Start(ctx, "Manager.Export")
+	defer exportSpan.End()
+	for {
+		select {
+		case fatal := <-taskErrors:
+			log.Errorw("fatal indexer error", "error", fatal)
+			return false, fatal
+		default:
+			res, ok := <-taskResults
+			if !ok {
+				log.Infow("index tipset complete", "height", ts.Height(), "success", success)
+				return success, nil
+			}
+
+			for _, report := range res.Report {
+				if report.Status != visormodel.ProcessingStatusOK &&
+					report.Status != visormodel.ProcessingStatusInfo {
+					log.Infow("task failed", "task", res.Name, "status", report.Status, "errors", report.ErrorsDetected, "info", report.StatusInformation)
+					success = false
+				} else {
+					log.Infow("task success", "task", res.Name, "status", report.Status, "duration", report.CompletedAt.Sub(report.StartedAt))
+				}
+			}
+			m := &ModelResult{
+				Name:  res.Name,
+				Model: model.PersistableList{res.Report, res.Data},
+			}
+
+			// synchronously export extracted data and its report.
```

Comment:
persisting `synchronously` here should be fine since the `taskResults` channel is buffered. Alternatively this could be performed with an errorGroup that we wait on for completion. I think there is a lot of value in this method `TipSet` returning once it's processed and persisted all results, otherwise, it's callers must go inspect the database to determine success.

Reply:
Agree and empirically we know that persistence is only a small fraction of the time spent actually indexing